### ISSUE TITLE
How FARM manage more than one answer in SquadProcessor

### ISSUE THREAD

**Issue by ftesser** [deepset-ai/FARM#768]
**Question**
Hello, I was looking to SquadProcessor `max_answers` parameter:

https://github.com/deepset-ai/FARM/blob/bec0a9aeecd696d94e42c72bef6e8b6788a381e0/farm/data_handler/processor.py#L1826

I understand that the idea is used this to manage the possibility of multiple answers (e.g. Squad dev set), but I do not understand how is working.

I see that here an array of that dimension is preallocated:
https://github.com/deepset-ai/FARM/blob/bec0a9aeecd696d94e42c72bef6e8b6788a381e0/farm/data_handler/processor.py#L2052

But then it seems that in the prediction_head is used just the first answer:
https://github.com/deepset-ai/FARM/blob/bec0a9aeecd696d94e42c72bef6e8b6788a381e0/farm/modeling/prediction_head.py#L1035

So how the others answers are used in the train?

**Comment by Timoeller** [comment]
Hey @ftesser thats quite a dig deep into our code :smile: 

And you also found the answer. For training we only use one answer, since SQuAD and NQ only have one answer in training. The multiple answers are just for n-way annotated dev and test sets.

Hope that makes sense. Do you need loss signals from multiple answers during training?

**Comment by ftesser** [comment]
Thanks for the feedback @Timoeller,  yes this makes sense to me and regarding the training with multiple answers (I am doing some fine-tuning on a specific domain dataset with multiple answers) I think that rather than propagating loss signals from multiple answers, a simple solution could be to manage that at json level duplicating the questions and adding the new answers. What do you think?

However something is not yet clear to me: how `max_answers` is used in inference (for example for dev and test set)?
For example, in the case of question with multiple answers of squad dev set,  label_idxs is filled with  multiple answers, and whats happen then?

Finally, because we have one answer in training, should be an good option to force `max_answers=1` during the training? Do you see any contraindication?

**Comment by Timoeller** [comment]
Glad I can help here.

> manage that at json level duplicating

Good Idea. Especially because working with multiple answers in one go would mean substantial changes to the modelling part.

> how max_answers is used in inference

I am a bit unsure about terminology here. max answers is only used in our Evaluator (where inference takes place). Our Inferencer for getting model predictions does not handle the labels, if there are any, and therefore does not use max answers at all.  

> force max_answers=1 during the training?

The max answers is ignored during training, so I would not worry about this value and the usage so much. The additional ram it uses is tiny compared to everything else. If you find it very unintuitive would you like to add some docstring instead of our `# todo explain how we only use first answer for train `? That would be very much appreciated.

**Comment by ftesser** [comment]
Many thanks for super-faster reply @Timoeller!

If I will have time I would like to investigate a little bit into the Evaluator part, but if Evaluator does not look into `label_idxs`, probably  `max_answers` is not necessary at all...

> If you find it very unintuitive would you like to add some docstring instead of our # todo explain how we only use first answer for train ? That would be very much appreciated.

Yes I will do it.

Thanks!

**Comment by Timoeller** [comment]
Ah sorry, I think you misread me?

Only the Evaluator uses max_answers and all label_idxs. The FARM class in infer.py "Inferencer" does not. 

The Evaluator is also inferencing of course, next to evaluating the outcome of the inference (this term inference is defined differently in some fields, for us inference is the process of making predictions based on input data).

**Comment by ftesser** [comment]
Thanks for the clarifications.

**Comment by Timoeller** [comment]
I think we can close the issue, right? Feel free to reopen!
